File: RangeSumOfBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        // Time Complexity: O(n)
        // Space Complexity: O(n)
        int result = 0;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            if(node == NULL) {
                continue;
            }
            if(low <= node->val && node->val <= high) {
                result += node->val;
            }
            q.push(node->left);
            q.push(node->right);
        }
        return result;
    }
};


// Breadth First Search
// Traverse all nodes --> Check for condition and add the values that fall within the range --> Return the sum

// Time Complexity: O(n)
// Space Complexity: O(n)

File: RangeSumOfBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        # Time Complexity: O(n)
        # Space Complexity: O(n)
        result = 0
        queue = [root]
        while queue:
            node = queue.pop(0)
            if not node:
                continue
            if low <= node.val <= high:
                result += node.val
            queue.append(node.left)
            queue.append(node.right)
        return result

# Breadth First Search
# Traverse all nodes --> Check for condition and add the values that fall within the range --> Return the sum

# Time Complexity: O(n)
# Space Complexity: O(n)
